Add option to disable federation ingress controller

kzwang opened the pull request:
**What this PR does / why we need it**:
Added an option to enable/disable federation ingress controller as currently federated ingresses doesn't work in environments other than GCE/GKE. Also ignore reconcile config maps if no federated ingresses exist.

**Which issue this PR fixes** 
fixes #33943

@quinton-hoole

**Release note**:
```release-note
Add `--controllers` flag to federation controller manager for enable/disable federation ingress controller
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36643)
<!-- Reviewable:end -->

kzwang committed:
Add option to disable federation ingress controller

nikhiljindal commented:
Thanks for sending this @kzwang 
@k8s-bot ok to test

cc @kubernetes/sig-cluster-federation

nikhiljindal commented on federation/cmd/federation-controller-manager/app/options/options.go:
```diff
@@ -115,5 +118,6 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&s.APIServerBurst, "federated-api-burst", s.APIServerBurst, "Burst to use while talking with federation apiserver")
 	fs.StringVar(&s.DnsProvider, "dns-provider", s.DnsProvider, "DNS provider. Valid values are: "+fmt.Sprintf("%q", dnsprovider.RegisteredDnsProviders()))
 	fs.StringVar(&s.DnsConfigFile, "dns-provider-config", s.DnsConfigFile, "Path to config file for configuring DNS provider.")
+	fs.BoolVar(&s.EnableIngressController, "enable-ingress-controller", s.EnableIngressController, "Enable federation ingress controller.")
```
This way we will need to add a flag for each resource.
How about making this a comma separated list of resources whose controllers should be disabled.
It will be empty by default. Users who do not want ingress controller can add "ingress" to that list.

kzwang replied:
How about adding the `runtime-config` option (which is available to api server) to controller manager as well for enable/disable specific resource?

nikhiljindal commented on federation/pkg/federation-controller/ingress/ingress_controller_test.go:
```diff
@@ -118,14 +118,14 @@ func TestIngressController(t *testing.T) {
 	t.Log("Adding Federated Ingress")
 	fedIngressWatch.Add(&ing1)
+
+	t.Log("Checking that UID annotation on Cluster 1 annotation was correctly updated after added Federated Ingress")
```
after added -> after adding

nikhiljindal commented:
Another possible solution that we have discussed in the past is that the controller manager can use apiserver's discovery API to figure out the resources that are enabled on the apiserver. Controller manager will then not instantiate the controller for a resource that is not enabled on the apiserver.
This way, there is a single place (the apiserver) to enable/disable resources.

With the solution in this PR, admin will have to ensure that ingress is enabled/disabled in both federation apiserver and federation controller manager and a mismatch can cause unexpected behaviors for users.

madhusudancs commented:
@nikhiljindal there are legitimate cases where users might want the API resource but not the controller, esp. when they want to supply their own controller implementation. I think it is useful to have an ability to just disable the controller. Do we solve this problem in kube-controller-manager at all today? If so, how?

kzwang commented:
@nikhiljindal I've made some changes to the PR.

- Changed the option to `runtime-config` (same as api server)
- Moved `resource_config.go` from `genericapiserver` to `util/config` as controller manager is not generic api server (I'm not sure `util/config` is the correct place or not)
- Use discovery API to check whether resource is enabled in API server or not. 

Ingress controller will be disabled if ingress resource is disabled in either controller manager or api server.

liggitt commented on pkg/genericapiserver/config.go:
```diff
@@ -55,6 +55,7 @@ import (
 	utilnet "k8s.io/kubernetes/pkg/util/net"
 	"k8s.io/kubernetes/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/version"
+	"k8s.io/kubernetes/pkg/util/config"
```
not super-excited about this as a package dependency... util/config seems super vague

kzwang replied:
@liggitt any suggestions on where it should be moved to?

deads2k commented:
I don't think this flag arrangement works very well.  `group/version/resource` isn't a boundary that works well for controllers.  Many (most?) controllers deal in multiple resources and several even span `group/versions`. For instance, the replica set controller, deployment controller, and job controller all fall into that category.  You can also imagine having multiple controllers working on the same logical resource (openshift build controllers are a good example).  I would rather see a logical categorization of controllers into buckets based on the functionality they provide and enable and disablement based on that.

I'd also like to see the ability to have "* minus" options.  By default, all controllers would be on.  If you specified, `--enable=jobs --enable=rs`, only jobs and RS would be on.  If you specify, `--disable=ingress`, everything except the ingress controller would be on.

This syntax is something I'd like to have in the general controller manager.  I just hadn't made it that far yet.

kzwang commented:
@deads2k Should controller manager also check api server to make sure the resource for specific controller is enabled? e.g. If user disable ingress or config map in api server but enable ingress controller this definitely won't work.

Also how about keep the runtime config syntax but instead of group/version/resource=true/false we change that to controller=true/false? Currently all controllers are enabled by default but it's possible later there are some alpha controllers are disabled by default and user can use this to enable those controllers.

deads2k commented:
> Also how about keep the runtime config syntax but instead of group/version/resource=true/false we change that to controller=true/false? Currently all controllers are enabled by default but it's possible later there are some alpha controllers are disabled by default and user can use this to enable those controllers.

The `group/version/resource` doesn't map cleanly to most controllers, so I don't think its a good fit for what we're trying to configure.



> @deads2k Should controller manager also check api server to make sure the resource for specific controller is enabled? e.g. If user disable ingress or config map in api server but enable ingress controller this definitely won't work.

If we go with an explicit enable/disable of a named area of controllers, then if they explicitly request the ingress controller, you should crash.  If they got by default (used disable rules only), then I think we should error (`utilruntime.HandleError`), but not crash.

kzwang commented:
@deads2k @nikhiljindal I've changed the flag to `controllers`, the value would be `controller=true/false`, e.g. `ingress=false`

By default, the ingress controller is enabled, if `igress=false` is set, ingress controller will be disabled.

For ingress controller, Ingress and ConfigMap resources need to be enabled in API server. If any of those resources are not enabled in API server, controller manager will:
- Log a warning message if `ingress=true` is not set explicitly
- Crash if `ingress=true` is set

nikhiljindal commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -53,6 +57,11 @@ const (
+)
+
+var (
+	RequiredResources = []unversioned.GroupVersionResource{extensionsapiv1beta1.SchemeGroupVersion.WithResource("ingresses"), apiv1.SchemeGroupVersion.WithResource("configmaps")}
```
Ingress controller does not require config maps to be enabled in federation-apiserver (It requires them to be present in underlying kube-apiservers).
config map was added to federation-apiserver after federated ingress controller was written.

nikhiljindal commented on federation/cmd/federation-controller-manager/app/options/options.go:
```diff
@@ -118,6 +119,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.DnsConfigFile, "dns-provider-config", s.DnsConfigFile, "Path to config file for configuring DNS provider.")
-	fs.BoolVar(&s.EnableIngressController, "enable-ingress-controller", s.EnableIngressController, "Enable federation ingress controller.")
+	fs.Var(&s.Controllers, "controllers", ""+
+		"A set of key=value pairs that describe controller configuration that may be passed "+
```
I wouldnt block this PR on that but will be great if we can list all valid values here. For ex: right now "ingress" is the only valid value. I can see people getting confused by should the key for ingress controller be "ingress" or "Ingress" or "ingress-controller" or "ingressController" or many more such combinations :)

nikhiljindal commented on federation/cmd/federation-controller-manager/app/options/options.go:
```diff
@@ -65,8 +66,8 @@ type ControllerManagerConfiguration struct {
 	ContentType string `json:"contentType"`
-	// enableIngressController enables the Federation Ingress Controller
-	EnableIngressController bool `json:"enableIngressController"`
+	// controllers configuration
```
ConfigMap determining which controllers should be enabled or disabled

nikhiljindal commented:
Thanks a lot @kzwang.
This LGTM. I just have a few minor comments.
Can you please squash your commits? Also add a release note for the new flag. I will fix the label.
 
Thx

kzwang commented:
@nikhiljindal I've squashed my commits and added the release note.

nikhiljindal commented:
Thanks @kzwang 
Fixed the release note label and added LGTM.

nikhiljindal commented:
@k8s-bot cvm gce e2e test this

kzwang commented:
@nikhiljindal I've rebased with the master branch, can you reapply the LGTM label?

rmmh commented:
@k8s-bot test this issue kubernetes/test-infra#1253

rmmh commented:
@k8s-bot test this issue kubernetes/test-infra#1253 (also, deleting spurious failure comments)

nikhiljindal commented:
Build is failing with the following error:
```
federation/cmd/federation-controller-manager/app/controllermanager.go:195: cannot use serverResources (type []*"k8s.io/kubernetes/pkg/apis/meta/v1".APIResourceList) as type map[string]*"k8s.io/kubernetes/pkg/apis/meta/v1".APIResourceList in argument to controllerEnabled
```

@kzwang Can you please take a look?

kzwang commented:
@nikhiljindal I've fixed the build failure.

nikhiljindal commented:
thx. added lgtm
